src/data/db.ts

```diff
@@ -0,0 +1,202 @@
+    );
+
+    if (!missingLanguages.length)
+      return existingLanguages.map((language) => language.id);
+
+    const columnsSet = new pgp.helpers.ColumnSet(["name", "slug"], {
+      table: "programming_language",
+    });
+
+    const missingRecords = missingLanguages.map((language) => {
+      return { name: language, slug: slugify(language) };
+    });
+
+    const query =
+      pgp.helpers.insert(missingRecords, columnsSet) +
+      " ON CONFLICT DO NOTHING RETURNING id";
+
+    return db.many(query).then((newLanguages) => {
+      return newLanguages
+        .map(({ id }) => id)
+        .concat(existingLanguages.map(({ id }) => id));
+    });
+  });
+}
+
+function linkUserAndLanguages(userId: number, languageIds: number[]) {
+  const records = languageIds.map((languageId) => {
+    return { user_id: userId, language_id: languageId };
+  });
+
+  const columnsSet = new pgp.helpers.ColumnSet(["user_id", "language_id"], {
+    table: "github_user_language",
+  });
+
+  const query =
+    pgp.helpers.insert(records, columnsSet) + " ON CONFLICT DO NOTHING";
+
+  return db.none(query);
+}
+
+function createGithubUser(
+  data: GithubUsers
+): PromiseLike<Pick<GithubUsers, "id">> {
+  return db
+    .one(
+      `
+      INSERT INTO
+        github_user (login, name, company, blog, email, location, bio, hireable)
+      VALUES
+        ($[login], $[name], $[company], $[blog], $[email], $[location], $[bio], $[hireable])
+      ON CONFLICT (login)
+        DO UPDATE SET name=$[name], company=$[company], blog=$[blog], email=$[email], location=$[location], bio=$[bio], hireable=$[hireable]
+      RETURNING id;
+    `,
+      data
+    )
+    .then((user) => {
+      if (!data.languages || !data.languages.length) return user;
+
+      return getLanguageIds(data.languages)
+        .then((languageIds) => linkUserAndLanguages(user.id, languageIds))
+        .then(() => user);
+    });
+}
+
+function findUserIdsWithLanguage(language: string) {
+  const query = `
+    SELECT
+      gul.user_id as id
+    FROM
+      github_user_language gul JOIN programming_language pl ON gul.language_id=pl.id
+    WHERE
+      pl.slug = $1;
+  `;
+
+  const promise = db.manyOrNone(query, [slugify(language)]);
+  return Promise.map(promise, ({ id }) => id);
+}
+
+function listUserLanguages(userId: number) {
+  const query = `
+    SELECT
+      pl.name
+    FROM
+      github_user_language gul JOIN programming_language pl ON gul.language_id=pl.id
+    WHERE
+      gul.user_id = $1;`;
+
+  const promise = db.manyOrNone(query, [userId]);
+  return Promise.map(promise, ({ name }) => name);
+}
+
+function listGithubUsers(
+  options: {
+    location?: string;
+    language?: string;
+    page?: number;
+    limit?: number;
+  } = {}
+) {
+  let promise = options.language
+    ? findUserIdsWithLanguage(options.language)
+    : Promise.resolve([]);
+
+  return promise.then((ids: number[]) => {
+    if (options.language && !ids.length) return [];
+
+    let query = [
```

Comment:
Maybe it would be clearer if you used [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) here

Reply:
Good call. There was a bug hidden here. If both filters were to be provided at the same time, the query would break due to the missing AND there 😓 .